java/common/src/main/java/org/ray/util/MethodId.java

```diff
@@ -0,0 +1,224 @@
+   * REF_putStatic               = 4,
+   * REF_invokeVirtual           = 5,
+   * REF_invokeStatic            = 6,
+   * REF_invokeSpecial           = 7,
+   * REF_newInvokeSpecial        = 8,
+   * REF_invokeInterface         = 9,
+   * REF_LIMIT                  = 10.
+   */
+
+  private static final ConcurrentHashMap<String, MethodId> MAP = new ConcurrentHashMap<>(256);
+  /**
+   * format A.B.C.cname
+   */
+  public final String className;
+  public final String methodName;
+  public final String methodDesc;
+  public final boolean isStatic;
+  private final String encoding;
+
+  private final byte[] digest;
+
+  private MethodId(String className, String methodName, String methodDesc, boolean isStatic,
+      String encoding) {
+    this.className = className;
+    this.methodName = methodName;
+    this.methodDesc = methodDesc;
+    this.isStatic = isStatic;
+    this.encoding = encoding;
+    this.digest = getSha1Hash0();
+  }
+
+  public MethodId(String className, String methodName, String methodDesc, boolean isStatic) {
+    this(className, methodName, methodDesc, isStatic,
+        encoding(className, methodName, methodDesc, isStatic));
+  }
+
+
+  private static String encoding(String className, String methodName, String methodDesc,
+      boolean isStatic) {
+    StringBuilder sb = new StringBuilder(512);
+    sb.append(className).append('/').append(methodName).append("::").append(methodDesc).append("&&")
+        .append(isStatic);
+    return sb.toString();
+  }
+
+  public static MethodId fromMethod(Method method, boolean forceNew) {
+    final boolean isstatic = Modifier.isStatic(method.getModifiers());
+    final String className = method.getDeclaringClass().getName();
+    final String methodName = method.getName();
+    final Type type = Type.getType(method);
+    final String methodDesc = type.getDescriptor();
+    final String encoding = encoding(className, methodName,
+        methodDesc, isstatic);
+    if (forceNew) {
+      return new MethodId(className, methodName,
+          methodDesc, isstatic, encoding);
+    }
+    MethodId m = MAP.get(encoding);
+    if (m == null) {
+      m = new MethodId(className, methodName,
+          methodDesc, isstatic, encoding);
+      MAP.putIfAbsent(encoding, m);
+      m = MAP.get(encoding);
+    }
+    return m;
+  }
+
+  public static MethodId fromSerializedLambda(SerializedLambda lambda) {
+    return fromSerializedLambda(lambda, false);
+  }
+
+  public static MethodId fromSerializedLambda(SerializedLambda lambda, boolean forceNew) {
+    if (lambda.getCapturedArgCount() != 0) {
+      throw new IllegalArgumentException("could not transfer a lambda which is closer");
+    }
+    //REF_invokeStatic
+    final boolean isstatic = lambda.getImplMethodKind() == 6;
+    final String className = lambda.getImplClass().replace('/', '.');
+    final String encoding = encoding(className, lambda.getImplMethodName(),
+        lambda.getImplMethodSignature(), isstatic);
```

Comment:
beacuse need the encoded string as a key to get the cache. there are 2 constructors, if the key has computed, not need to do that again in constructor

Reply:
I think we can do either `cache.get(m.encoding)`, or simply use MethodId as the key of cache (because MethodId's hashCode is just its encoding).
Then we only need one constructor. Does it look cleaner?

Reply:
Ok, I will do that